### PR TITLE
Add at_midBlock support

### PR DESCRIPTION
**Pull request by PepperCode1**
- Capture local position when rendering blocks to calculate midBlock
- Rename sodium.impl.block_id package to sodium.impl.block_context
- Rename MaterialIdHolder to BlockContextHolder and move to
sodium.impl.block_context
- Rename MaterialIdAwareVertexWriter to ContextAwareVertexWriter
- Remove mc_Entity attribute from extended entity format
- Move mc_Entity render type magic values to constant field

### CODE REVIEW

File: src/sodiumCompatibility/java/net/coderbot/iris/compat/sodium/mixin/vertex_format/MixinMultidrawChunkRenderBackend.java [hunk 1]
```diff
@@ -31,14 +31,16 @@ public MixinMultidrawChunkRenderBackend(ChunkVertexType vertexType) {
 					ordinal = 0))
 	private GlVertexAttributeBinding[] iris$addAdditionalBindings(GlVertexAttributeBinding[] base) {
 		return BlockRenderingSettings.INSTANCE.shouldUseExtendedVertexFormat() ? ArrayUtils.addAll(base,
-				new GlVertexAttributeBinding(IrisChunkShaderBindingPoints.BLOCK_ID,
-						vertexFormat.getAttribute(IrisChunkMeshAttributes.BLOCK_ID)),
-				new GlVertexAttributeBinding(IrisChunkShaderBindingPoints.MID_TEX_COORD,
-						vertexFormat.getAttribute(IrisChunkMeshAttributes.MID_TEX_COORD)),
+				new GlVertexAttributeBinding(IrisChunkShaderBindingPoints.NORMAL,
```
Comment (IMS212) [inline src/sodiumCompatibility/java/net/coderbot/iris/compat/sodium/mixin/vertex_format/MixinMultidrawChunkRenderBackend.java:34]: Any reason for shuffling these?

File: src/main/java/net/coderbot/iris/mixin/vertices/MixinBufferBuilder.java [hunk 2]
```diff
@@ -135,14 +152,28 @@ public abstract class MixinBufferBuilder implements BufferVertexConsumer, BlockS
...
 		this.nextElement();
+		if (iris$isTerrain) {
+			// MID_BLOCK_ELEMENT
+			int posIndex = this.nextElementByte - format.getVertexSize() + 4;
+			float x = buffer.getFloat(posIndex);
+			float y = buffer.getFloat(posIndex + 4);
+			float z = buffer.getFloat(posIndex + 8);
+			this.putInt(0, ExtendedDataHelper.computeMidBlock(x, y, z, currentLocalPosX, currentLocalPosY, currentLocalPosZ));
+			this.nextElement();
+		}
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/mixin/vertices/MixinBufferBuilder.java:176]: This code seems to silently assume that MID_BLOCK_ELEMENT is the last element. In addition, would we want to handle text (which uses the TERRAIN vertex format) differently?

**Review by coderbot16 — COMMENTED** [review]
Great work! One small comment:

**Comment by coderbot16** [comment]
Also, there appears to be conflicts with the other PR merges.

**Review by coderbot16 — APPROVED** [review]
Thanks!

**Pull request merged**